sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -81,13 +84,157 @@ private DoFnSignatures() {}
     return signature;
   }
 
+  /**
+   * The global context for the class, for use in analysis.
+   *
+   * <p>It contains much of the information that eventually becomes part of the {@link
+   * DoFnSignature}, but in an intermediate state.
+   */
+  @AutoValue
+  abstract static class GlobalAnalysisContext {
```

Comment:
I'm not convinced that these Context classes need to be immutable, or even use AutoValue. We use them in a mutable fashion ("context = context.withBlah()"), because they capture the changing state of analysis, and they don't escape the current class, so it seems easier to make them mutable and reduce the boilerplate.